### PR TITLE
Add forksafe threading objects

### PR DESCRIPTION
**Pull request by jd**
## refactor(forksafe): add forksafe threading Lock classes

This adds a wrapper around the Lock, RLock and Event class in order to provide
a fork-safe equivalent of the existing classes.

This objects can be re-used after a fork safely as they are reset.

## refactor(profiling): leverage forksafe.Lock in DoubleLock

This leverages the forksafe Lock to prevent dead-lock after fork.
This makes sure that when the parent-Profiler is stopped in the child process,
it can actually do it without trouble since the recorder DoubleLock is sure to
be freed.

## fix(service): use forksafe Lock and Event objects

This replaces the existing and regular Lock and Event classes from the Python
threading by their forksafe alternative, making sure that a Service can be
safely stopped or a PeriodicThread joined once a process has forked.

### CODE REVIEW

File: ddtrace/internal/nogevent.py [hunk 1]
```diff
@@ -44,8 +45,8 @@ def is_module_patched(module):
     class DoubleLock(object):
         """A lock that prevent concurrency from a gevent coroutine and from a threading.Thread at the same time."""
 
-        _lock = attr.ib(factory=threading.Lock, init=False, repr=False)
-        _thread_lock = attr.ib(factory=Lock, init=False, repr=False)
+        _lock = attr.ib(factory=forksafe.Lock, init=False, repr=False)
+        _thread_lock = attr.ib(factory=lambda: forksafe.ResetObject(Lock), init=False, repr=False)
```
Comment (brettlangdon) [inline ddtrace/internal/nogevent.py:49]:
```suggestion
        _thread_lock = attr.ib(factory=forksafe.Lock, init=False, repr=False)
```

?

File: ddtrace/internal/forksafe.py [hunk 2]
```diff
@@ -55,3 +57,55 @@ def _after_fork():
...
+            log.exception("Exception ignored in object reset forksafe hook %r", obj)
+
+
+register(_reset_objects)
+
+
+class ResetObject(wrapt.ObjectProxy):
+    """An object wrapper object that is fork-safe and resets itself after a fork.
+
+    When a Python process forks, a Lock can be in any state, locked or not, by any thread. Since after after fork all
```
Comment (brettlangdon) [inline ddtrace/internal/forksafe.py:80]: "after after fork" ?

mistyped and have double "after", or it is meant to be ready as 'after "after fork"' ?

File: tests/tracer/test_forksafe.py [hunk 3]
```diff
@@ -146,3 +150,39 @@ def f3():
...
+def test_lock_basic():
+    # type: (...) -> None
+    lock = forksafe.Lock()
+    assert lock.acquire()
+    assert lock.release() is None
+    with pytest.raises(lock_release_exc_type):
+        lock.release()
+
+
+def test_lock_fork():
```
Comment (brettlangdon) [inline tests/tracer/test_forksafe.py:170]: docstrings on these tests would be nice

**Review by brettlangdon — COMMENTED** [review]
are there any test cases we should add around our internal behaviors to ensure this works as expected?

that is, if we know the runtime metrics thread can cause these issues, is it possible to simulate that in an integration test and assert this change fixes that?

**Comment by brettlangdon** [comment]
I think this should have a changelog

File: ddtrace/internal/forksafe.py (hunk 2 above)
Comment (jd) [reply ddtrace/internal/forksafe.py:80]: mistyped :)

File: ddtrace/internal/nogevent.py (hunk 1 above)
Comment (jd) [reply ddtrace/internal/nogevent.py:49]: Ok, I'll add a comment :)

Comment (brettlangdon) [reply ddtrace/internal/nogevent.py:51]: oh! I totally misread/misunderstood `Lock` vs `threading.Lock` usage here, makes sense!

**Review by brettlangdon — COMMENTED** [review]
lgtm

**Review by P403n1x87 — APPROVED** [review]
shame we need to implement this stuff ourselves 🙁

**Review by brettlangdon — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]
lgtm, but I still think a generic changelog entry would be worthwhile.

e.g. "Improve fork safety of blah blah blah"

what do you think?

**Comment by jd** [comment]
It's internal _cuisine_, I'm not sure it provides value to users to read through that. 🤷

**Comment by brettlangdon** [comment]
> It's internal _cuisine_, I'm not sure it provides value to users to read through that. 🤷

fair... my only thinking is if someone is having issues with multi-threaded applications and they scan to look for improvements to that?

but maybe a little digging deep on reason to add. not a strong opinion

File: ddtrace/internal/forksafe.py [hunk 4]
```diff
@@ -55,3 +57,55 @@ def _after_fork():
...
+            obj._reset_object()
+        except Exception:
+            log.exception("Exception ignored in object reset forksafe hook %r", obj)
+
+
+register(_reset_objects)
+
+
+class ResetObject(wrapt.ObjectProxy):
+    """An object wrapper object that is fork-safe and resets itself after a fork.
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/forksafe.py:78]: It's not really clear what "resets" means here. Might be worth saying "reinstantiate" to be direct?

File: ddtrace/internal/forksafe.py [hunk 5]
```diff
@@ -55,3 +57,55 @@ def _after_fork():
...
+    def __init__(
+        self, wrapped_class  # type: typing.Any
+    ):
+        # type: (...) -> None
+        super(ResetObject, self).__init__(wrapped_class())
+        self._self_wrapped_class = wrapped_class
+        _resetable_objects.add(self)
+
+    def _reset_object(self):
+        # type: (...) -> None
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/forksafe.py:95]: I wonder if we should have some kind of interface such that objects can control their recreation if needed (ex: `TracerWriter` defines a `recreate` API which is to essentially handle reinitializing after a fork). Might be handy if some attributes need to be persisted.

Definitely not required for now and we can add it if/when necessary but wanted to brain dump here if there are any future readers.

**Review by Kyle-Verhoog — DISMISSED** [review]
🍴 🦺 

Just one suggestion and an idea for the future but otherwise looks great.

I attached the manual merge label since @majorgreys is a requested reviewer still 🙂

File: ddtrace/internal/forksafe.py [hunk 6]
```diff
@@ -55,3 +57,55 @@ def _after_fork():
...
+_resetable_objects = weakref.WeakSet()  # type: weakref.WeakSet[ResetObject]
+
+
+def _reset_objects():
+    # type: (...) -> None
+    for obj in _resetable_objects:
+        try:
+            obj._reset_object()
+        except Exception:
+            log.exception("Exception ignored in object reset forksafe hook %r", obj)
```
Comment (P403n1x87) [inline ddtrace/internal/forksafe.py:71]:
```suggestion
            log.exception("Exception ignored in object reset forksafe hook %r", obj, exc_info=True)
```
?

Comment (jd) [reply ddtrace/internal/forksafe.py:71]: `log.exception` 😉

Comment (P403n1x87) [reply ddtrace/internal/forksafe.py:71]: Ah sorry! In my mind that read `error` 😊

File: tests/tracer/test_forksafe.py [hunk 7]
```diff
@@ -146,3 +150,44 @@ def f3():
...
     assert exit_code == 12
+
+
+if six.PY2:
+    lock_release_exc_type = _thread.error
+else:
+    lock_release_exc_type = RuntimeError
+
+
+def test_lock_basic():
```
Comment (majorgreys) [inline tests/tracer/test_forksafe.py:161]: Can we add similar tests for `RLock` and `Event`?

File: ddtrace/internal/forksafe.py (hunk 5 above)
Comment (majorgreys) [reply ddtrace/internal/forksafe.py:95]: https://github.com/DataDog/dd-trace-py/pull/2245 will help for this though is there something other than the threading that you have in mind here @Kyle-Verhoog ?

File: tests/tracer/test_forksafe.py (hunk 7 above)
Comment (jd) [reply tests/tracer/test_forksafe.py:161]: Will do!

File: ddtrace/internal/forksafe.py (hunk 5 above)
Comment (Kyle-Verhoog) [reply ddtrace/internal/forksafe.py:95]: Oh yes, looks like that would do it

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**